# Curso de pandas
## Primera parte
## Creando un Series con diccionario

In [1]:
# Importamos la libreria pandas

import pandas as pd

sdata = {"caracas":20000,"tachira":12000,"barinas":52000,"vargas":37000}

In [19]:
obj1 = pd.Series(sdata)
obj1

caracas    20000
tachira    12000
barinas    52000
vargas     37000
dtype: int64

In [29]:
# Se pueden cambiar los indices (estados) de la siguiente manera:

estados = ["caracas","tachira","falcon","vargas"]
obj2 = pd.Series(sdata, index=estados)
obj2

caracas    20000.0
tachira    12000.0
falcon         NaN
vargas     37000.0
dtype: float64

In [30]:
# Se puede ver que falcon es un valor faltante o vacio y para saber si hay valores faltantes
# usamos el metodo pd.isnull() devuelve un valor booleano.

pd.isnull(obj2)

caracas    False
tachira    False
falcon      True
vargas     False
dtype: bool

In [31]:
# Por el contrario los valores que no estan vacios pd.notnull()

pd.notnull(obj2)

caracas     True
tachira     True
falcon     False
vargas      True
dtype: bool

In [32]:
# Se pueden usar los metodos con los nombres de los objetos, por ejemplo:

obj2.isnull()

caracas    False
tachira    False
falcon      True
vargas     False
dtype: bool

In [17]:
obj1.notnull()

caracas    True
tachira    True
barinas    True
vargas     True
dtype: bool

In [ ]:
# Se pueden sumar objetos

In [33]:
obj1 + obj2

barinas        NaN
caracas    40000.0
falcon         NaN
tachira    24000.0
vargas     74000.0
dtype: float64

# Creando Data Frame

In [6]:
datos = {"states":["Ohio","Ohio","Ohio","Nevada","Nevada","Nevada"],
         "year":[2000,2001,2002,2001,2002,2003],
         "pop":[1.5,1.7,3.6,2.4,2.9,3.2]}

In [7]:
# Para convertir el diccionario en un dataframe usamos el metodo pd.DataFrame()

df = pd.DataFrame(datos)
df

states  year  pop
0    Ohio  2000  1.5
1    Ohio  2001  1.7
2    Ohio  2002  3.6
3  Nevada  2001  2.4
4  Nevada  2002  2.9
5  Nevada  2003  3.2

In [8]:
# Para saber cuales son los primeros registros usamos el metodo nombre_del_objeto.head(cantidad de registros)

df.head(3)

states  year  pop
0   Ohio  2000  1.5
1   Ohio  2001  1.7
2   Ohio  2002  3.6

In [61]:
# Para saber cuales son los ultimos registros usamos el metodo nombre_del_objeto.head.tail(cantidad de registros)

df.tail(4)

states  year  pop
2    Ohio  2002  3.6
3  Nevada  2001  2.4
4  Nevada  2002  2.9
5  Nevada  2003  3.2

In [38]:
# Para reordenar o cambiar el orden de las columnas se puede usar

pd.DataFrame(datos, columns = ["year","states","pop"])

year  states  pop
0  2000    Ohio  1.5
1  2001    Ohio  1.7
2  2002    Ohio  3.6
3  2001  Nevada  2.4
4  2002  Nevada  2.9
5  2003  Nevada  3.2

In [9]:
# Para cambiar los indices de las filas

df2=pd.DataFrame(datos, columns=["states","year","pop","debt"], index=["one","two","three","four","five","six"])
df2

states  year  pop debt
one      Ohio  2000  1.5  NaN
two      Ohio  2001  1.7  NaN
three    Ohio  2002  3.6  NaN
four   Nevada  2001  2.4  NaN
five   Nevada  2002  2.9  NaN
six    Nevada  2003  3.2  NaN

In [22]:
# Se puede seleccionar columnas especificas con el nombre de la siguiente manera

df2["year"]

one      2000
two      2001
three    2002
four     2001
five     2002
six      2003
Name: year, dtype: int64

In [43]:
# Si queremos rellenar la columna debt

df2["debt"] = 2.16
df2

states  year  pop  debt
one      Ohio  2000  1.5  2.16
two      Ohio  2001  1.7  2.16
three    Ohio  2002  3.6  2.16
four   Nevada  2001  2.4  2.16
five   Nevada  2002  2.9  2.16
six    Nevada  2003  3.2  2.16

## Importar archivos csv a python
Vamos a importar un archivo csv que contiene todos los datos de jugadores de fifa 2020

In [52]:
player_20 = pd.read_csv("../Jupyter Notebook/Datasets/archive/players_20.csv")
player_20.head(5)

sofifa_id                                         player_url  \
0     158023  https://sofifa.com/player/158023/lionel-messi/...   
1      20801  https://sofifa.com/player/20801/c-ronaldo-dos-...   
2     190871  https://sofifa.com/player/190871/neymar-da-sil...   
3     200389  https://sofifa.com/player/200389/jan-oblak/20/...   
4     183277  https://sofifa.com/player/183277/eden-hazard/2...   

          short_name                            long_name  age         dob  \
0           L. Messi       Lionel Andrés Messi Cuccittini   32  1987-06-24   
1  Cristiano Ronaldo  Cristiano Ronaldo dos Santos Aveiro   34  1985-02-05   
2          Neymar Jr        Neymar da Silva Santos Junior   27  1992-02-05   
3           J. Oblak                            Jan Oblak   26  1993-01-07   
4          E. Hazard                          Eden Hazard   28  1991-01-07   

   height_cm  weight_kg nationality                 club  ...   lwb   ldm  \
0        170         72   Argentina         FC Barcelona  ...  68+2  66+2   
1        187         83    Portugal             Juventus  ...  65+3  61+3   
2        175         68      Brazil  Paris Saint-Germain  ...  66+3  61+3   
3        188         87    Slovenia      Atlético Madrid  ...   NaN   NaN   
4        175         74     Belgium          Real Madrid  ...  66+3  63+3   

    cdm   rdm   rwb    lb   lcb    cb   rcb    rb  
0  66+2  66+2  68+2  63+2  52+2  52+2  52+2  63+2  
1  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3  
2  61+3  61+3  66+3  61+3  46+3  46+3  46+3  61+3  
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
4  63+3  63+3  66+3  61+3  49+3  49+3  49+3  61+3  

[5 rows x 104 columns]

## Analisis preliminar de datos

In [3]:
# El siguiente dataset es de recursos humanos de una empresa

dfrh = pd.read_csv("../Jupyter Notebook/Datasets/Recursos_Humanos/HRDataset_v14.csv")
dfrh.head(3)

Employee_Name  EmpID  MarriedID  MaritalStatusID  GenderID  \
0       Adinolfi, Wilson  K  10026          0                0         1   
1  Ait Sidi, Karthikeyan     10084          1                1         1   
2         Akinkuolie, Sarah  10196          1                1         0   

   EmpStatusID  DeptID  PerfScoreID  FromDiversityJobFairID  Salary  ...  \
0            1       5            4                       0   62506  ...   
1            5       3            3                       0  104437  ...   
2            5       5            3                       0   64955  ...   

      ManagerName  ManagerID RecruitmentSource PerformanceScore  \
0  Michael Albert       22.0          LinkedIn          Exceeds   
1      Simon Roup        4.0            Indeed      Fully Meets   
2  Kissy Sullivan       20.0          LinkedIn      Fully Meets   

   EngagementSurvey EmpSatisfaction SpecialProjectsCount  \
0              4.60               5                    0   
1              4.96               3                    6   
2              3.02               3                    0   

  LastPerformanceReview_Date DaysLateLast30 Absences  
0                  1/17/2019              0        1  
1                  2/24/2016              0       17  
2                  5/15/2012              0        3  

[3 rows x 36 columns]

In [58]:
# Hay que conocer la forma de nuestro dataset, es decir, que cantidad de filas y de columnas tenemos con el metodo .shape

dfrh.shape

(311, 36)

In [59]:
# Para saber cuales son los tipos de cada una de las variables o columnas usamos el metodo .dtypes

dfrh.dtypes

Employee_Name                  object
EmpID                           int64
MarriedID                       int64
MaritalStatusID                 int64
GenderID                        int64
EmpStatusID                     int64
DeptID                          int64
PerfScoreID                     int64
FromDiversityJobFairID          int64
Salary                          int64
Termd                           int64
PositionID                      int64
Position                       object
State                          object
Zip                             int64
DOB                            object
Sex                            object
MaritalDesc                    object
CitizenDesc                    object
HispanicLatino                 object
RaceDesc                       object
DateofHire                     object
DateofTermination              object
TermReason                     object
EmploymentStatus               object
Department                     object
ManagerName 

### Para seleccionar solamente columnas especificas que nos interesen se debe crear otro dataframe e igualarlo al dataframe principal especificando las columnas entre doble corchetes [[]]

In [6]:
dfrh1 = dfrh[["Employee_Name","Salary","Position","State","Sex","DOB","MaritalDesc","CitizenDesc","Department",
              "PerformanceScore","EngagementSurvey","EmpSatisfaction"]]
dfrh1.head(4)

#Notemos que antes teniamos 36 columnas y ahora solo tenemos 12

Employee_Name  Salary                  Position State Sex  \
0       Adinolfi, Wilson  K   62506   Production Technician I    MA  M    
1  Ait Sidi, Karthikeyan     104437                   Sr. DBA    MA  M    
2         Akinkuolie, Sarah   64955  Production Technician II    MA   F   
3              Alagbe,Trina   64991   Production Technician I    MA   F   

        DOB MaritalDesc CitizenDesc         Department PerformanceScore  \
0  07/10/83      Single  US Citizen  Production                 Exceeds   
1  05/05/75     Married  US Citizen              IT/IS      Fully Meets   
2  09/19/88     Married  US Citizen  Production             Fully Meets   
3  09/27/88     Married  US Citizen  Production             Fully Meets   

   EngagementSurvey  EmpSatisfaction  
0              4.60                5  
1              4.96                3  
2              3.02                3  
3              4.84                5

## Segunda parte
## Datos faltantes en un dataset
Esta parte la trabajaremos con datos de personas del titanic, importaremos un archivo excel

In [16]:
# Importamos el archivo excel

df_titanic = pd.read_excel("../Jupyter Notebook/Datasets/Titanic/titanic3.xls")

### Exploracion

In [45]:
# Desplegamos las primeras filas para conocer un poco el dataset

df_titanic.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

       age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.0000      0      0   24160  211.3375       B5        S    2    NaN   
1   0.9167      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.0000      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

In [20]:
# Podemos usar la funcion describe() para desplegar las medidas de tendencia central y las medidas
# de dispercion de las columnas cuantitativas

df_titanic.describe()

pclass     survived          age        sibsp        parch  \
count  1309.000000  1309.000000  1046.000000  1309.000000  1309.000000   
mean      2.294882     0.381971    29.881135     0.498854     0.385027   
std       0.837836     0.486055    14.413500     1.041658     0.865560   
min       1.000000     0.000000     0.166700     0.000000     0.000000   
25%       2.000000     0.000000    21.000000     0.000000     0.000000   
50%       3.000000     0.000000    28.000000     0.000000     0.000000   
75%       3.000000     1.000000    39.000000     1.000000     0.000000   
max       3.000000     1.000000    80.000000     8.000000     9.000000   

              fare        body  
count  1308.000000  121.000000  
mean     33.295479  160.809917  
std      51.758668   97.696922  
min       0.000000    1.000000  
25%       7.895800   72.000000  
50%      14.454200  155.000000  
75%      31.275000  256.000000  
max     512.329200  328.000000

La función describe() de Pandas regresa estadísticas descriptivas incluyendo: media, meadiana, máx, mín, std y conteos para una columna en particular de los datos. La función describe solo regresa los valores de estas estadísticas para las columnas numéricas.

In [28]:
df_titanic.shape

(1309, 14)

### Eliminar filas con NAN

In [32]:
titanic1 = df_titanic.dropna(axis=0, how="all")

# El parametro axis selecciona filas = 0 o columnas = 1
# Con how="all" le estamos diciendo que elimine todas las filas que tienen TODOS los valores faltantes

In [33]:
titanic1.shape

(1309, 14)

Como podemos ver este dataset se mantiene exactanemnet igual al dataset original, esto quiere decir que no hay ninguna fila que tenga todos sus valores faltantes

In [37]:
titanic2 = df_titanic.dropna(axis=0, how="any")

# Con how="any" le estamos diciendo que elimine todas las filas que tenga cualquier (al menos 1) valor faltante

In [36]:
titanic2.shape

(0, 14)

Esto quiere decir que de las 1309 filas, todas tenian por lo menos un objeto faltante

### Rellenar NAN
Vamos a analizar la columna de "age"

In [41]:
# Este codigo lo que hace es contar los valores faltantes en la columna "age"

pd.isnull(df_titanic["age"]).values.ravel().sum()

# La funcion pd.isnull() devuelve los valores nulos de la columna "age", despues los valores y la sumatoria

263

Como vemos de los 1309 registros de la columna "age" 263 son nulos

In [88]:
# Vamos a rellenar esos 263 age (edades) faltantes

col = df_titanic["age"].fillna(df_titanic["age"].mean())

# Con df_titanic["age"] seleccionamos la columna
# Con el metodo .fillna() rellenamos los valores nulos con lo que haya en el argumento
# En el argumento esta df_titanic["age"].mean() que saca el promedio de esa columna
# En resumen rellenamos los espacios vacios de la columna "age" con el promedio (mean) de edades

col.loc[12:16]

12    24.000000
13    26.000000
14    80.000000
15    29.881135
16    24.000000
Name: age, dtype: float64

In [93]:
# Vamos a crear una nueva columna en el dataframe de titanic1 para no modificar el original

titanic1["new_age"] = col
titanic1.loc[14:16]

pclass  survived                                  name   sex   age  sibsp  \
14       1         1  Barkworth, Mr. Algernon Henry Wilson  male  80.0      0   
15       1         0                   Baumann, Mr. John D  male   NaN      0   
16       1         0              Baxter, Mr. Quigg Edmond  male  24.0      0   

    parch    ticket      fare    cabin embarked boat  body      home.dest  \
14      0     27042   30.0000      A23        S    B   NaN  Hessle, Yorks   
15      0  PC 17318   25.9250      NaN        S  NaN   NaN   New York, NY   
16      1  PC 17558  247.5208  B58 B60        C  NaN   NaN   Montreal, PQ   

      new_age  
14  80.000000  
15  29.881135  
16  24.000000

Como podemos ver en la fila 15, en la columna "age" hay un vacio y se relleno en la columna "new_age" con el promedio de edades de todos los pasajeros

In [96]:
# Ahora vamos a eliminar la columna vieja de "age"

titanic1 = titanic1[["name","sex","new_age","sibsp","parch","ticket","fare","cabin","embarked","boat","body","home.dest"]]
titanic1.head(3)

name     sex  new_age  sibsp  parch  ticket  \
0   Allen, Miss. Elisabeth Walton  female  29.0000      0      0   24160   
1  Allison, Master. Hudson Trevor    male   0.9167      1      2  113781   
2    Allison, Miss. Helen Loraine  female   2.0000      1      2  113781   

       fare    cabin embarked boat  body                        home.dest  
0  211.3375       B5        S    2   NaN                     St Louis, MO  
1  151.5500  C22 C26        S   11   NaN  Montreal, PQ / Chesterville, ON  
2  151.5500  C22 C26        S  NaN   NaN  Montreal, PQ / Chesterville, ON

## One hot encoding
Usaremos el dataset de Recursos Humanos

In [31]:
dfrh = pd.read_csv("../Jupyter Notebook/Datasets/Recursos_Humanos/HRDataset_v14.csv")
dfrh.head()

Employee_Name  EmpID  MarriedID  MaritalStatusID  GenderID  \
0       Adinolfi, Wilson  K  10026          0                0         1   
1  Ait Sidi, Karthikeyan     10084          1                1         1   
2         Akinkuolie, Sarah  10196          1                1         0   
3              Alagbe,Trina  10088          1                1         0   
4          Anderson, Carol   10069          0                2         0   

   EmpStatusID  DeptID  PerfScoreID  FromDiversityJobFairID  Salary  ...  \
0            1       5            4                       0   62506  ...   
1            5       3            3                       0  104437  ...   
2            5       5            3                       0   64955  ...   
3            1       5            3                       0   64991  ...   
4            5       5            3                       0   50825  ...   

      ManagerName  ManagerID RecruitmentSource PerformanceScore  \
0  Michael Albert       22.0          LinkedIn          Exceeds   
1      Simon Roup        4.0            Indeed      Fully Meets   
2  Kissy Sullivan       20.0          LinkedIn      Fully Meets   
3    Elijiah Gray       16.0            Indeed      Fully Meets   
4  Webster Butler       39.0     Google Search      Fully Meets   

   EngagementSurvey EmpSatisfaction SpecialProjectsCount  \
0              4.60               5                    0   
1              4.96               3                    6   
2              3.02               3                    0   
3              4.84               5                    0   
4              5.00               4                    0   

  LastPerformanceReview_Date DaysLateLast30 Absences  
0                  1/17/2019              0        1  
1                  2/24/2016              0       17  
2                  5/15/2012              0        3  
3                   1/3/2019              0       15  
4                   2/1/2016              0        2  

[5 rows x 36 columns]

In [32]:
# Hacemos una seleccion de columnas para hacer un analisis un poco mas enfocado

dfrh1 = dfrh[["Employee_Name","Salary","Position","State","Sex","DOB","MaritalDesc","CitizenDesc",
              "Department","PerformanceScore","EngagementSurvey","EmpSatisfaction"]]
dfrh1.head()

Employee_Name  Salary                  Position State Sex  \
0       Adinolfi, Wilson  K   62506   Production Technician I    MA  M    
1  Ait Sidi, Karthikeyan     104437                   Sr. DBA    MA  M    
2         Akinkuolie, Sarah   64955  Production Technician II    MA   F   
3              Alagbe,Trina   64991   Production Technician I    MA   F   
4          Anderson, Carol    50825   Production Technician I    MA   F   

        DOB MaritalDesc CitizenDesc         Department PerformanceScore  \
0  07/10/83      Single  US Citizen  Production                 Exceeds   
1  05/05/75     Married  US Citizen              IT/IS      Fully Meets   
2  09/19/88     Married  US Citizen  Production             Fully Meets   
3  09/27/88     Married  US Citizen  Production             Fully Meets   
4  09/08/89    Divorced  US Citizen  Production             Fully Meets   

   EngagementSurvey  EmpSatisfaction  
0              4.60                5  
1              4.96                3  
2              3.02                3  
3              4.84                5  
4              5.00                4

In [33]:
dfrh1.shape

(311, 12)

In [20]:
dfrh1.describe()

Salary  EngagementSurvey  EmpSatisfaction
count     311.000000        311.000000       311.000000
mean    69020.684887          4.110000         3.890675
std     25156.636930          0.789938         0.909241
min     45046.000000          1.120000         1.000000
25%     55501.500000          3.690000         3.000000
50%     62810.000000          4.280000         4.000000
75%     72036.000000          4.700000         5.000000
max    250000.000000          5.000000         5.000000

In [34]:
# Usamos el one hot encoding con la columna de sexo, que crear dos columnas: masculino y femenino
# Creamos un objeto 

dummy_sex = pd.get_dummies(dfrh1["Sex"], prefix="sexo")

In [35]:
dummy_sex.head()

sexo_F  sexo_M 
0       0        1
1       0        1
2       1        0
3       1        0
4       1        0

In [36]:
# Vamos a eliminar la columna de sex

dfrh1 = dfrh1.drop(["Sex"], axis = 1)

In [37]:
dfrh1.shape

(311, 11)

In [38]:
dfrh1.head(3)

Employee_Name  Salary                  Position State       DOB  \
0       Adinolfi, Wilson  K   62506   Production Technician I    MA  07/10/83   
1  Ait Sidi, Karthikeyan     104437                   Sr. DBA    MA  05/05/75   
2         Akinkuolie, Sarah   64955  Production Technician II    MA  09/19/88   

  MaritalDesc CitizenDesc         Department PerformanceScore  \
0      Single  US Citizen  Production                 Exceeds   
1     Married  US Citizen              IT/IS      Fully Meets   
2     Married  US Citizen  Production             Fully Meets   

   EngagementSurvey  EmpSatisfaction  
0              4.60                5  
1              4.96                3  
2              3.02                3

In [40]:
# Ahora vamos a agregar las columnas de sex (dummy_sex)
# Creamos un nuevo dataset y le concatenamos el anterior con las columnas nuevas

dfrh2 = pd.concat([dfrh1, dummy_sex], axis = 1)

In [42]:
dfrh2.shape

(311, 13)

In [43]:
dfrh2.head(4)

Employee_Name  Salary                  Position State       DOB  \
0       Adinolfi, Wilson  K   62506   Production Technician I    MA  07/10/83   
1  Ait Sidi, Karthikeyan     104437                   Sr. DBA    MA  05/05/75   
2         Akinkuolie, Sarah   64955  Production Technician II    MA  09/19/88   
3              Alagbe,Trina   64991   Production Technician I    MA  09/27/88   

  MaritalDesc CitizenDesc         Department PerformanceScore  \
0      Single  US Citizen  Production                 Exceeds   
1     Married  US Citizen              IT/IS      Fully Meets   
2     Married  US Citizen  Production             Fully Meets   
3     Married  US Citizen  Production             Fully Meets   

   EngagementSurvey  EmpSatisfaction  sexo_F  sexo_M   
0              4.60                5       0        1  
1              4.96                3       0        1  
2              3.02                3       1        0  
3              4.84                5       1        0

In [52]:
# Ahora vamos a automatizar el mismo proceso con una funcion tomando como ejemplo CitizenDesc

def createdummies(df, varname): # Creamos una funcion, los parametros son el dataframe y la columna a aplicar one hot encoding
    dummy=pd.get_dummies(df[varname], prefix = varname) # Convertimos la variable categorica a variable numerica o contable
    df=df.drop([varname], axis = 1) # Eliminamos la columna de la variable categorica
    df=pd.concat([df, dummy], axis = 1) # Concatenamos el dataframe con las culumnas de la variable contable
    return df.head(4)

In [53]:
createdummies(dfrh2, "CitizenDesc")

Employee_Name  Salary                  Position State       DOB  \
0       Adinolfi, Wilson  K   62506   Production Technician I    MA  07/10/83   
1  Ait Sidi, Karthikeyan     104437                   Sr. DBA    MA  05/05/75   
2         Akinkuolie, Sarah   64955  Production Technician II    MA  09/19/88   
3              Alagbe,Trina   64991   Production Technician I    MA  09/27/88   

  MaritalDesc         Department PerformanceScore  EngagementSurvey  \
0      Single  Production                 Exceeds              4.60   
1     Married              IT/IS      Fully Meets              4.96   
2     Married  Production             Fully Meets              3.02   
3     Married  Production             Fully Meets              4.84   

   EmpSatisfaction  sexo_F  sexo_M   CitizenDesc_Eligible NonCitizen  \
0                5       0        1                                0   
1                3       0        1                                0   
2                3       1        0                                0   
3                5       1        0                                0   

   CitizenDesc_Non-Citizen  CitizenDesc_US Citizen  
0                        0                       1  
1                        0                       1  
2                        0                       1  
3                        0                       1

## Como concatenar en pandas


In [4]:
dfrh.head()

Employee_Name  EmpID  MarriedID  MaritalStatusID  GenderID  \
0       Adinolfi, Wilson  K  10026          0                0         1   
1  Ait Sidi, Karthikeyan     10084          1                1         1   
2         Akinkuolie, Sarah  10196          1                1         0   
3              Alagbe,Trina  10088          1                1         0   
4          Anderson, Carol   10069          0                2         0   

   EmpStatusID  DeptID  PerfScoreID  FromDiversityJobFairID  Salary  ...  \
0            1       5            4                       0   62506  ...   
1            5       3            3                       0  104437  ...   
2            5       5            3                       0   64955  ...   
3            1       5            3                       0   64991  ...   
4            5       5            3                       0   50825  ...   

      ManagerName  ManagerID RecruitmentSource PerformanceScore  \
0  Michael Albert       22.0          LinkedIn          Exceeds   
1      Simon Roup        4.0            Indeed      Fully Meets   
2  Kissy Sullivan       20.0          LinkedIn      Fully Meets   
3    Elijiah Gray       16.0            Indeed      Fully Meets   
4  Webster Butler       39.0     Google Search      Fully Meets   

   EngagementSurvey EmpSatisfaction SpecialProjectsCount  \
0              4.60               5                    0   
1              4.96               3                    6   
2              3.02               3                    0   
3              4.84               5                    0   
4              5.00               4                    0   

  LastPerformanceReview_Date DaysLateLast30 Absences  
0                  1/17/2019              0        1  
1                  2/24/2016              0       17  
2                  5/15/2012              0        3  
3                   1/3/2019              0       15  
4                   2/1/2016              0        2  

[5 rows x 36 columns]

In [5]:
# Creamos un nuevo dataset con solo 4 columnas y 6 filas

dfrh_prueba = dfrh[["Employee_Name","Salary","PerformanceScore","EngagementSurvey"]] #Creamos las 4 columnas
dfrh_prueba = dfrh_prueba.head(6) # Creamos las 6 filas
dfrh_prueba

Employee_Name  Salary PerformanceScore  EngagementSurvey
0       Adinolfi, Wilson  K   62506          Exceeds              4.60
1  Ait Sidi, Karthikeyan     104437      Fully Meets              4.96
2         Akinkuolie, Sarah   64955      Fully Meets              3.02
3              Alagbe,Trina   64991      Fully Meets              4.84
4          Anderson, Carol    50825      Fully Meets              5.00
5         Anderson, Linda     57568          Exceeds              5.00

In [6]:
# Creamos otro dataset con la misma informacion pero diferente estructura

dfrh_prueba1 = dfrh[["Employee_Name","Salary","PerformanceScore","EngagementSurvey"]]
dfrh_prueba1 = dfrh_prueba1.loc[7:12]
dfrh_prueba1

Employee_Name  Salary PerformanceScore  EngagementSurvey
7            Athwal, Sam   59365      Fully Meets              5.00
8       Bachiochi, Linda   47837      Fully Meets              4.46
9     Bacong, Alejandro    50178      Fully Meets              5.00
10  Baczenski, Rachael     54670      Fully Meets              4.20
11       Barbara, Thomas   47211      Fully Meets              4.20
12      Barbossa, Hector   92328          Exceeds              4.28

In [8]:
# Ahora vamos a concatenar o unir los dos datasets

pd.concat([dfrh_prueba,dfrh_prueba1])

# Concatenamos dos tablas con las mismas columnas

Employee_Name  Salary PerformanceScore  EngagementSurvey
0        Adinolfi, Wilson  K   62506          Exceeds              4.60
1   Ait Sidi, Karthikeyan     104437      Fully Meets              4.96
2          Akinkuolie, Sarah   64955      Fully Meets              3.02
3               Alagbe,Trina   64991      Fully Meets              4.84
4           Anderson, Carol    50825      Fully Meets              5.00
5          Anderson, Linda     57568          Exceeds              5.00
7                Athwal, Sam   59365      Fully Meets              5.00
8           Bachiochi, Linda   47837      Fully Meets              4.46
9         Bacong, Alejandro    50178      Fully Meets              5.00
10      Baczenski, Rachael     54670      Fully Meets              4.20
11           Barbara, Thomas   47211      Fully Meets              4.20
12          Barbossa, Hector   92328          Exceeds              4.28

In [9]:
# Creamos una nueva tabla con la concatenacion de las dos anteriores

dfrh_prueba2 = pd.concat([dfrh_prueba,dfrh_prueba1])

In [10]:
# La tabla anterior la vamos a separar en dos tablas:
# Una tabla de una columna

dfrh_prueba3 = dfrh_prueba2[["Employee_Name"]]

In [11]:
# Una tabla de tres columnas

dfrh_prueba4 = dfrh_prueba2[["Salary","Salary","EngagementSurvey"]]

In [12]:
dfrh_prueba3

Employee_Name
0        Adinolfi, Wilson  K
1   Ait Sidi, Karthikeyan   
2          Akinkuolie, Sarah
3               Alagbe,Trina
4           Anderson, Carol 
5          Anderson, Linda  
7                Athwal, Sam
8           Bachiochi, Linda
9         Bacong, Alejandro 
10      Baczenski, Rachael  
11           Barbara, Thomas
12          Barbossa, Hector

In [13]:
dfrh_prueba4

Salary  Salary  EngagementSurvey
0    62506   62506              4.60
1   104437  104437              4.96
2    64955   64955              3.02
3    64991   64991              4.84
4    50825   50825              5.00
5    57568   57568              5.00
7    59365   59365              5.00
8    47837   47837              4.46
9    50178   50178              5.00
10   54670   54670              4.20
11   47211   47211              4.20
12   92328   92328              4.28

In [15]:
# El metodo anterior era para concatenar filas aqui vamos a concatenar las dos columnas nuevamente

pd.concat([dfrh_prueba3,dfrh_prueba4], axis=1)

Employee_Name  Salary  Salary  EngagementSurvey
0        Adinolfi, Wilson  K   62506   62506              4.60
1   Ait Sidi, Karthikeyan     104437  104437              4.96
2          Akinkuolie, Sarah   64955   64955              3.02
3               Alagbe,Trina   64991   64991              4.84
4           Anderson, Carol    50825   50825              5.00
5          Anderson, Linda     57568   57568              5.00
7                Athwal, Sam   59365   59365              5.00
8           Bachiochi, Linda   47837   47837              4.46
9         Bacong, Alejandro    50178   50178              5.00
10      Baczenski, Rachael     54670   54670              4.20
11           Barbara, Thomas   47211   47211              4.20
12          Barbossa, Hector   92328   92328              4.28

## Tercera parte
## Sort Values - Rename
### Ordenamiento de valores

In [16]:
# Importamos el dataset de fifa20

fifa20 = pd.read_csv("../Jupyter Notebook/Datasets/Players_fifa20/players_20.csv")
fifa20.head()

sofifa_id                                         player_url  \
0     158023  https://sofifa.com/player/158023/lionel-messi/...   
1      20801  https://sofifa.com/player/20801/c-ronaldo-dos-...   
2     190871  https://sofifa.com/player/190871/neymar-da-sil...   
3     200389  https://sofifa.com/player/200389/jan-oblak/20/...   
4     183277  https://sofifa.com/player/183277/eden-hazard/2...   

          short_name                            long_name  age         dob  \
0           L. Messi       Lionel Andrés Messi Cuccittini   32  1987-06-24   
1  Cristiano Ronaldo  Cristiano Ronaldo dos Santos Aveiro   34  1985-02-05   
2          Neymar Jr        Neymar da Silva Santos Junior   27  1992-02-05   
3           J. Oblak                            Jan Oblak   26  1993-01-07   
4          E. Hazard                          Eden Hazard   28  1991-01-07   

   height_cm  weight_kg nationality                 club  ...   lwb   ldm  \
0        170         72   Argentina         FC Barcelona  ...  68+2  66+2   
1        187         83    Portugal             Juventus  ...  65+3  61+3   
2        175         68      Brazil  Paris Saint-Germain  ...  66+3  61+3   
3        188         87    Slovenia      Atlético Madrid  ...   NaN   NaN   
4        175         74     Belgium          Real Madrid  ...  66+3  63+3   

    cdm   rdm   rwb    lb   lcb    cb   rcb    rb  
0  66+2  66+2  68+2  63+2  52+2  52+2  52+2  63+2  
1  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3  
2  61+3  61+3  66+3  61+3  46+3  46+3  46+3  61+3  
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
4  63+3  63+3  66+3  61+3  49+3  49+3  49+3  61+3  

[5 rows x 104 columns]

In [17]:
fifa20.tail()

sofifa_id                                         player_url  \
18273     245006  https://sofifa.com/player/245006/shuai-shao/20...   
18274     250995  https://sofifa.com/player/250995/mingjie-xiao/...   
18275     252332  https://sofifa.com/player/252332/wei-zhang/20/...   
18276     251110  https://sofifa.com/player/251110/haijian-wang/...   
18277     233449  https://sofifa.com/player/233449/ximing-pan/20...   

         short_name     long_name  age         dob  height_cm  weight_kg  \
18273    Shao Shuai            邵帅   22  1997-03-10        186         79   
18274  Xiao Mingjie  Mingjie Xiao   22  1997-01-01        177         66   
18275     Zhang Wei            张威   19  2000-05-16        186         75   
18276  Wang Haijian           汪海健   18  2000-08-02        185         74   
18277    Pan Ximing           潘喜明   26  1993-01-11        182         78   

      nationality                           club  ...   lwb   ldm   cdm   rdm  \
18273    China PR               Beijing Renhe FC  ...  43+2  42+2  42+2  42+2   
18274    China PR               Shanghai SIPG FC  ...  44+2  43+2  43+2  43+2   
18275    China PR         Hebei China Fortune FC  ...  47+2  49+2  49+2  49+2   
18276    China PR  Shanghai Greenland Shenhua FC  ...  48+2  48+2  48+2  48+2   
18277    China PR         Hebei China Fortune FC  ...  48+2  49+2  49+2  49+2   

        rwb    lb   lcb    cb   rcb    rb  
18273  43+2  45+2  46+2  46+2  46+2  45+2  
18274  44+2  46+2  47+2  47+2  47+2  46+2  
18275  47+2  47+2  49+2  49+2  49+2  47+2  
18276  48+2  48+2  49+2  49+2  49+2  48+2  
18277  48+2  48+2  50+2  50+2  50+2  48+2  

[5 rows x 104 columns]

In [18]:
fifa20.shape

(18278, 104)

In [19]:
# Creamos un nuevo dataset con menos columnas

fifa20_1 = fifa20[["short_name","age","height_cm","weight_kg","nationality","club","overall","potential","value_eur",
                   "wage_eur","player_positions","power_jumping","power_long_shots"]]

In [21]:
fifa20_1.head()

short_name  age  height_cm  weight_kg nationality  \
0           L. Messi   32        170         72   Argentina   
1  Cristiano Ronaldo   34        187         83    Portugal   
2          Neymar Jr   27        175         68      Brazil   
3           J. Oblak   26        188         87    Slovenia   
4          E. Hazard   28        175         74     Belgium   

                  club  overall  potential  value_eur  wage_eur  \
0         FC Barcelona       94         94   95500000    565000   
1             Juventus       93         93   58500000    405000   
2  Paris Saint-Germain       92         92  105500000    290000   
3      Atlético Madrid       91         93   77500000    125000   
4          Real Madrid       91         91   90000000    470000   

  player_positions  power_jumping  power_long_shots  
0       RW, CF, ST             68                94  
1           ST, LW             95                93  
2          LW, CAM             61                84  
3               GK             78                12  
4           LW, CF             56                80

In [22]:
fifa20_1.shape

(18278, 13)

### Ordenar de menor a mayor

In [28]:
#Para ordenar las columnas utilizamos el metodo sort_values("Columna")

fifa20_1.sort_values("overall").head()

short_name  age  height_cm  weight_kg          nationality  \
18277          Pan Ximing   26        182         78             China PR   
18252         A. Aksungur   18        175         72               Turkey   
18251  G. Sykes-Kenworthy   19        182         78              England   
18250           R. Roache   19        178         74  Republic of Ireland   
18249       Shan Huanhuan   20        185         75             China PR   

                         club  overall  potential  value_eur  wage_eur  \
18277  Hebei China Fortune FC       48         51      40000      2000   
18252             Denizlispor       48         65      60000      1000   
18251           Bradford City       48         65      50000      1000   
18250               Blackpool       48         65      60000      1000   
18249        Dalian YiFang FC       48         65      70000      2000   

      player_positions  power_jumping  power_long_shots  
18277               CM             64                30  
18252               CM             61                30  
18251               GK             52                 8  
18250               ST             68                41  
18249               ST             34                39

### Ordenar de mayor a menor

In [32]:
# En el parametro de sort_values() le agregamos ascending=False

fifa20_1.sort_values("overall", ascending=False).head()

short_name  age  height_cm  weight_kg nationality  \
0           L. Messi   32        170         72   Argentina   
1  Cristiano Ronaldo   34        187         83    Portugal   
2          Neymar Jr   27        175         68      Brazil   
3           J. Oblak   26        188         87    Slovenia   
4          E. Hazard   28        175         74     Belgium   

                  club  overall  potential  value_eur  wage_eur  \
0         FC Barcelona       94         94   95500000    565000   
1             Juventus       93         93   58500000    405000   
2  Paris Saint-Germain       92         92  105500000    290000   
3      Atlético Madrid       91         93   77500000    125000   
4          Real Madrid       91         91   90000000    470000   

  player_positions  power_jumping  power_long_shots  
0       RW, CF, ST             68                94  
1           ST, LW             95                93  
2          LW, CAM             61                84  
3               GK             78                12  
4           LW, CF             56                80

In [35]:
# Ahora vamos a crear un dataset pero solo con los jugadores de un solo equipo
# Vamos a filtrar a los jugadores del FC Barcelona

fcBarcelona = fifa20_1[fifa20_1["club"] == "FC Barcelona"]
fcBarcelona.head()

short_name  age  height_cm  weight_kg nationality          club  \
0          L. Messi   32        170         72   Argentina  FC Barcelona   
6     M. ter Stegen   27        187         85     Germany  FC Barcelona   
19        L. Suárez   32        182         86     Uruguay  FC Barcelona   
21  Sergio Busquets   30        189         76       Spain  FC Barcelona   
22     A. Griezmann   28        176         73      France  FC Barcelona   

    overall  potential  value_eur  wage_eur player_positions  power_jumping  \
0        94         94   95500000    565000       RW, CF, ST             68   
6        90         93   67500000    250000               GK             79   
19       89         89   53000000    355000               ST             69   
21       89         89   55000000    300000          CDM, CM             66   
22       89         89   69000000    370000       CF, ST, LW             89   

    power_long_shots  
0                 94  
6                 10  
19                86  
21                54  
22                83

In [39]:
# Vamos a ordenar a los jugadores por el overall de mayor a menor

fcBarcelona.sort_values("overall", ascending=False).head()

short_name  age  height_cm  weight_kg nationality          club  \
0          L. Messi   32        170         72   Argentina  FC Barcelona   
6     M. ter Stegen   27        187         85     Germany  FC Barcelona   
19        L. Suárez   32        182         86     Uruguay  FC Barcelona   
21  Sergio Busquets   30        189         76       Spain  FC Barcelona   
22     A. Griezmann   28        176         73      France  FC Barcelona   

    overall  potential  value_eur  wage_eur player_positions  power_jumping  \
0        94         94   95500000    565000       RW, CF, ST             68   
6        90         93   67500000    250000               GK             79   
19       89         89   53000000    355000               ST             69   
21       89         89   55000000    300000          CDM, CM             66   
22       89         89   69000000    370000       CF, ST, LW             89   

    power_long_shots  
0                 94  
6                 10  
19                86  
21                54  
22                83

In [42]:
# Vamos a ver a los jugadores con mayor valoracion en euros

fcBarcelona.sort_values("value_eur", ascending=False).head()

short_name  age  height_cm  weight_kg nationality          club  \
0          L. Messi   32        170         72   Argentina  FC Barcelona   
22     A. Griezmann   28        176         73      France  FC Barcelona   
6     M. ter Stegen   27        187         85     Germany  FC Barcelona   
21  Sergio Busquets   30        189         76       Spain  FC Barcelona   
19        L. Suárez   32        182         86     Uruguay  FC Barcelona   

    overall  potential  value_eur  wage_eur player_positions  power_jumping  \
0        94         94   95500000    565000       RW, CF, ST             68   
22       89         89   69000000    370000       CF, ST, LW             89   
6        90         93   67500000    250000               GK             79   
21       89         89   55000000    300000          CDM, CM             66   
19       89         89   53000000    355000               ST             69   

    power_long_shots  
0                 94  
22                83  
6                 10  
21                54  
19                86

### Cambiar nombre de columnas

In [46]:
# Usamos las funcion rename(columns="before_name":"after_name")

fcBarcelona.rename(columns = {"age":"edad","value_eur":"valor_eur"}).head()

short_name  edad  height_cm  weight_kg nationality          club  \
0          L. Messi    32        170         72   Argentina  FC Barcelona   
6     M. ter Stegen    27        187         85     Germany  FC Barcelona   
19        L. Suárez    32        182         86     Uruguay  FC Barcelona   
21  Sergio Busquets    30        189         76       Spain  FC Barcelona   
22     A. Griezmann    28        176         73      France  FC Barcelona   

    overall  potential  valor_eur  wage_eur player_positions  power_jumping  \
0        94         94   95500000    565000       RW, CF, ST             68   
6        90         93   67500000    250000               GK             79   
19       89         89   53000000    355000               ST             69   
21       89         89   55000000    300000          CDM, CM             66   
22       89         89   69000000    370000       CF, ST, LW             89   

    power_long_shots  
0                 94  
6                 10  
19                86  
21                54  
22                83

### Como eliminar columnas

In [47]:
#  Con la funcion df.drop(columns=["Length","Height"])

fcb = fcBarcelona.drop(columns=["short_name","age"])

In [48]:
fcb.head()

height_cm  weight_kg nationality          club  overall  potential  \
0         170         72   Argentina  FC Barcelona       94         94   
6         187         85     Germany  FC Barcelona       90         93   
19        182         86     Uruguay  FC Barcelona       89         89   
21        189         76       Spain  FC Barcelona       89         89   
22        176         73      France  FC Barcelona       89         89   

    value_eur  wage_eur player_positions  power_jumping  power_long_shots  
0    95500000    565000       RW, CF, ST             68                94  
6    67500000    250000               GK             79                10  
19   53000000    355000               ST             69                86  
21   55000000    300000          CDM, CM             66                54  
22   69000000    370000       CF, ST, LW             89                83

### Mostrar filas que cumplan un criterio

In [49]:
# Vamos a filtrar los jugadores del Fc Barcelona que tengan un overall mayor a 88

fcBarcelona[fcBarcelona.overall > 88]

short_name  age  height_cm  weight_kg nationality          club  \
0          L. Messi   32        170         72   Argentina  FC Barcelona   
6     M. ter Stegen   27        187         85     Germany  FC Barcelona   
19        L. Suárez   32        182         86     Uruguay  FC Barcelona   
21  Sergio Busquets   30        189         76       Spain  FC Barcelona   
22     A. Griezmann   28        176         73      France  FC Barcelona   

    overall  potential  value_eur  wage_eur player_positions  power_jumping  \
0        94         94   95500000    565000       RW, CF, ST             68   
6        90         93   67500000    250000               GK             79   
19       89         89   53000000    355000               ST             69   
21       89         89   55000000    300000          CDM, CM             66   
22       89         89   69000000    370000       CF, ST, LW             89   

    power_long_shots  
0                 94  
6                 10  
19                86  
21                54  
22                83

In [50]:
# Vamos a filtrar los jugadores que tengan una edad mayor a 30

fcBarcelona[fcBarcelona.age > 30]

short_name  age  height_cm  weight_kg nationality          club  overall  \
0      L. Messi   32        170         72   Argentina  FC Barcelona       94   
19    L. Suárez   32        182         86     Uruguay  FC Barcelona       89   
29        Piqué   32        194         85       Spain  FC Barcelona       88   
64   I. Rakitić   31        184         78     Croatia  FC Barcelona       86   
142    A. Vidal   32        180         75       Chile  FC Barcelona       84   

     potential  value_eur  wage_eur player_positions  power_jumping  \
0           94   95500000    565000       RW, CF, ST             68   
19          89   53000000    355000               ST             69   
29          88   38000000    285000               CB             75   
64          86   38000000    245000          CM, CDM             34   
142         84   23500000    205000          CM, CDM             85   

     power_long_shots  
0                  94  
19                 86  
29                 51  
64                 85  
142                82

In [55]:
# Vamos a filtrar los jugadores que tengan un peso mayor a 82

fcBarcelona[fcBarcelona.weight_kg > 82]

short_name  age  height_cm  weight_kg nationality          club  \
6   M. ter Stegen   27        187         85     Germany  FC Barcelona   
19      L. Suárez   32        182         86     Uruguay  FC Barcelona   
29          Piqué   32        194         85       Spain  FC Barcelona   
91           Neto   29        190         84      Brazil  FC Barcelona   

    overall  potential  value_eur  wage_eur player_positions  power_jumping  \
6        90         93   67500000    250000               GK             79   
19       89         89   53000000    355000               ST             69   
29       88         88   38000000    285000               CB             75   
91       85         86   31000000    175000               GK             83   

    power_long_shots  
6                 10  
19                86  
29                51  
91                20

In [56]:
fcBarcelona[fcBarcelona.weight_kg > 82].sort_values("weight_kg")

short_name  age  height_cm  weight_kg nationality          club  \
91           Neto   29        190         84      Brazil  FC Barcelona   
6   M. ter Stegen   27        187         85     Germany  FC Barcelona   
29          Piqué   32        194         85       Spain  FC Barcelona   
19      L. Suárez   32        182         86     Uruguay  FC Barcelona   

    overall  potential  value_eur  wage_eur player_positions  power_jumping  \
91       85         86   31000000    175000               GK             83   
6        90         93   67500000    250000               GK             79   
29       88         88   38000000    285000               CB             75   
19       89         89   53000000    355000               ST             69   

    power_long_shots  
91                20  
6                 10  
29                51  
19                86

### Seleccion aleatorea de filas (fraccion)

In [57]:
fcBarcelona.shape

(33, 13)

In [58]:
fcBarcelona.head()

short_name  age  height_cm  weight_kg nationality          club  \
0          L. Messi   32        170         72   Argentina  FC Barcelona   
6     M. ter Stegen   27        187         85     Germany  FC Barcelona   
19        L. Suárez   32        182         86     Uruguay  FC Barcelona   
21  Sergio Busquets   30        189         76       Spain  FC Barcelona   
22     A. Griezmann   28        176         73      France  FC Barcelona   

    overall  potential  value_eur  wage_eur player_positions  power_jumping  \
0        94         94   95500000    565000       RW, CF, ST             68   
6        90         93   67500000    250000               GK             79   
19       89         89   53000000    355000               ST             69   
21       89         89   55000000    300000          CDM, CM             66   
22       89         89   69000000    370000       CF, ST, LW             89   

    power_long_shots  
0                 94  
6                 10  
19                86  
21                54  
22                83

In [60]:
# Aqui haremos el filtrado a travez de una fraccion con la funcion df.sample(frac=0.1)   sample = muestra
# En frac ponemos el porcentaje de filas que se mostraran de forma aleatorea, en este caso el 10% de las 33 filas

fcBarcelona.sample(frac=0.1)

short_name  age  height_cm  weight_kg  nationality          club  \
249   Nélson Semedo   25        177         67     Portugal  FC Barcelona   
4781        L. Reis   19        178         75  Netherlands  FC Barcelona   
4780         H. Abe   20        171         65        Japan  FC Barcelona   

      overall  potential  value_eur  wage_eur player_positions  power_jumping  \
249        82         87   26000000    150000               RB             80   
4781       70         83    3100000     29000          CDM, CM             82   
4780       70         83    3400000     44000           LM, RM             68   

      power_long_shots  
249                 60  
4781                60  
4780                55

### Seleccion aleatorea de filas (numeros)
En este caso ya no es porcentual sino de manera absoluta

In [61]:
# Aqui estamos seleccionando 5 filas de forma aleatorea

fcBarcelona.sample(n=5)

short_name  age  height_cm  weight_kg nationality          club  \
11042      Iñaki Peña   20        184         78       Spain  FC Barcelona   
7712     Carles Pérez   21        172         65       Spain  FC Barcelona   
7711   Oriol Busquets   20        185         77       Spain  FC Barcelona   
59          S. Umtiti   25        182         75      France  FC Barcelona   
110        O. Dembélé   22        178         67      France  FC Barcelona   

       overall  potential  value_eur  wage_eur player_positions  \
11042       64         81     850000     12000               GK   
7712        67         82    1700000     27000           RM, RW   
7711        67         82    1600000     24000          CDM, CM   
59          86         89   50000000    210000               CB   
110         84         90   42500000    195000           RW, LW   

       power_jumping  power_long_shots  
11042             63                 8  
7712              60                68  
7711              52                47  
59                88                74  
110               59                78

## Cuarta parte
### Seleccion de filas por posicion

In [2]:
# Cargamos los datos en un data frame de marketing

df_marketing = pd.read_csv("Datasets/Marketing/marketing_data.csv")
df_marketing.head()

ID  Year_Birth   Education Marital_Status      Income   Kidhome  \
0   1826        1970  Graduation       Divorced  $84,835.00         0   
1      1        1961  Graduation         Single  $57,091.00         0   
2  10476        1958  Graduation        Married  $67,267.00         0   
3   1386        1967  Graduation       Together  $32,474.00         1   
4   5371        1989  Graduation         Single  $21,474.00         1   

   Teenhome Dt_Customer  Recency  MntWines  ...  NumStorePurchases  \
0         0     6/16/14        0       189  ...                  6   
1         0     6/15/14        0       464  ...                  7   
2         1     5/13/14        0       134  ...                  5   
3         1     5/11/14        0        10  ...                  2   
4         0      4/8/14        0         6  ...                  2   

   NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  \
0                  1             0             0             0             0   
1                  5             0             0             0             0   
2                  2             0             0             0             0   
3                  7             0             0             0             0   
4                  7             1             0             0             0   

   AcceptedCmp2  Response  Complain  Country  
0             0         1         0       SP  
1             1         1         0       CA  
2             0         0         0       US  
3             0         0         0      AUS  
4             0         1         0       SP  

[5 rows x 28 columns]

In [4]:
df_marketing.shape

(2240, 28)

In [6]:
# Seleccion de filas de la 10 a la 20, no incluye la 20

df_marketing.iloc[10:20]

ID  Year_Birth   Education Marital_Status      Income   Kidhome  \
10   2079        1947    2n Cycle        Married  $81,044.00         0   
11   5642        1979      Master       Together  $62,499.00         1   
12  10530        1959         PhD          Widow  $67,786.00         0   
13   2964        1981  Graduation        Married  $26,872.00         0   
14  10311        1969  Graduation        Married   $4,428.00         0   
15    837        1977  Graduation        Married  $54,809.00         1   
16  10521        1977  Graduation        Married  $54,809.00         1   
17  10175        1958         PhD       Divorced  $32,173.00         0   
18   1473        1960    2n Cycle         Single  $47,823.00         0   
19   2795        1958      Master         Single  $30,523.00         2   

    Teenhome Dt_Customer  Recency  MntWines  ...  NumStorePurchases  \
10         0    12/27/13        0       450  ...                 10   
11         0     12/9/13        0       140  ...                  6   
12         0     12/7/13        0       431  ...                  6   
13         0    10/16/13        0         3  ...                  2   
14         1     10/5/13        0        16  ...                  0   
15         1     9/11/13        0        63  ...                  5   
16         1     9/11/13        0        63  ...                  5   
17         1      8/1/13        0        18  ...                  3   
18         1     7/23/13        0        53  ...                  3   
19         1      7/1/13        0         5  ...                  2   

    NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  \
10                  1             0             0             0             0   
11                  4             0             0             0             0   
12                  1             0             0             0             0   
13                  6             0             0             0             0   
14                  1             0             0             0             0   
15                  4             0             0             0             0   
16                  4             0             0             0             0   
17                  4             0             0             0             0   
18                  8             0             0             0             0   
19                  7             0             0             0             0   

    AcceptedCmp2  Response  Complain  Country  
10             0         0         0       US  
11             0         0         0       SP  
12             0         1         0      IND  
13             0         0         0       CA  
14             0         0         0       SP  
15             0         0         0       SP  
16             0         1         0       SP  
17             0         0         0       SP  
18             0         0         0       CA  
19             0         0         0       CA  

[10 rows x 28 columns]

In [25]:
# Para saber el nombre de cada columna usamos el metodo .columns.values

df_marketing.columns.values

array(['ID', 'Year_Birth', 'Education', 'Marital_Status', ' Income ',
       'Kidhome', 'Teenhome', 'Dt_Customer', 'Recency', 'MntWines',
       'MntFruits', 'MntMeatProducts', 'MntFishProducts',
       'MntSweetProducts', 'MntGoldProds', 'NumDealsPurchases',
       'NumWebPurchases', 'NumCatalogPurchases', 'NumStorePurchases',
       'NumWebVisitsMonth', 'AcceptedCmp3', 'AcceptedCmp4',
       'AcceptedCmp5', 'AcceptedCmp1', 'AcceptedCmp2', 'Response',
       'Complain', 'Country'], dtype=object)

### Otro tipo de seleccion de filas
¿Los clientes mas jovenes?

In [26]:
# Se utiliza el siguiente codigo para encontrar los 5 valores mayores de la columna "Year_Birth", con esto encontraremos
# los clientes mas jovenes

df_marketing.nlargest(5, "Year_Birth")

ID  Year_Birth   Education Marital_Status      Income   Kidhome  \
562    9909        1996    2n Cycle        Married   $7,500.00         0   
1824    193        1996       Basic        Married  $14,421.00         0   
697   10548        1995  Graduation         Single  $71,163.00         0   
964    3661        1995    2n Cycle         Single  $80,617.00         0   
1298   4427        1995    2n Cycle         Single  $83,257.00         0   

      Teenhome Dt_Customer  Recency  MntWines  ...  NumStorePurchases  \
562          0     11/9/12       24         3  ...                  3   
1824         0     2/17/14       81         0  ...                  2   
697          0      3/9/14       30       283  ...                 12   
964          0    10/12/12       42       594  ...                  8   
1298         0     9/18/12       56       536  ...                 12   

      NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  \
562                   9             0             0             0   
1824                  5             1             0             0   
697                   1             0             0             0   
964                   2             0             0             0   
1298                  6             1             0             1   

      AcceptedCmp1  AcceptedCmp2  Response  Complain  Country  
562              0             0         1         0      AUS  
1824             0             0         0         0       SP  
697              0             0         0         0      GER  
964              0             0         0         0      IND  
1298             0             0         1         1       SP  

[5 rows x 28 columns]

¿Los clientes mayores?

In [28]:
# Se utiliza el siguiente codigo para encontrar los 5 valores menores de la columna "Year_Birth", con esto encontraremos
# los clientes mas viejos o con mas edad

df_marketing.nsmallest(5, "Year_Birth")

ID  Year_Birth Education Marital_Status      Income   Kidhome  \
513   11004        1893  2n Cycle         Single  $60,182.00         0   
827    1150        1899       PhD       Together  $83,532.00         0   
2233   7829        1900  2n Cycle       Divorced  $36,640.00         1   
2171   6663        1940       PhD         Single  $51,141.00         0   
1740   6932        1941       PhD        Married  $93,027.00         0   

      Teenhome Dt_Customer  Recency  MntWines  ...  NumStorePurchases  \
513          1     5/17/14       23         8  ...                  2   
827          0     9/26/13       36       755  ...                  4   
2233         0     9/26/13       99        15  ...                  2   
2171         0      7/8/13       96       144  ...                  4   
1740         0     4/13/13       77      1285  ...                  5   

      NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  \
513                   4             0             0             0   
827                   1             0             0             1   
2233                  5             0             0             0   
2171                  5             0             0             0   
1740                  2             0             0             1   

      AcceptedCmp1  AcceptedCmp2  Response  Complain  Country  
513              0             0         0         0       SA  
827              0             0         0         0       SP  
2233             0             0         0         1      IND  
2171             0             0         0         0       SP  
1740             0             0         0         0       SP  

[5 rows x 28 columns]

¿Clientes con la mayor cantidad de dias desde la ultima compra?

In [32]:
# Con esto podemos encontrar a los clientes con los valores mas altos en la columna de "Recensy", es decir, se ubica
# a los clientes con la mayor cantidad de dias desde la ultima compra

df_marketing.nlargest(5, "Recency")

ID  Year_Birth   Education Marital_Status      Income   Kidhome  \
2223  2831        1976  Graduation       Together  $78,416.00         0   
2224   868        1966  Graduation        Married  $44,794.00         0   
2225  7212        1966  Graduation        Married  $44,794.00         0   
2226  1743        1974  Graduation         Single  $69,719.00         0   
2227  2415        1962  Graduation       Together  $62,568.00         0   

      Teenhome Dt_Customer  Recency  MntWines  ...  NumStorePurchases  \
2223         1     6/27/14       99       453  ...                 10   
2224         1      6/8/14       99        54  ...                  3   
2225         1      6/8/14       99        54  ...                  3   
2226         0     5/26/14       99       273  ...                  5   
2227         1      4/7/14       99       362  ...                  5   

      NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  \
2223                  3             0             0             0   
2224                  6             0             0             0   
2225                  6             0             0             0   
2226                  1             0             0             0   
2227                  4             0             0             0   

      AcceptedCmp1  AcceptedCmp2  Response  Complain  Country  
2223             0             0         0         0       SP  
2224             0             0         0         0       SA  
2225             0             0         0         0      AUS  
2226             0             0         0         0       SP  
2227             1             0         0         0       SP  

[5 rows x 28 columns]

¿Clientes con la menor cantidad de dias desde la ultima compra?

In [34]:
# Con esto podemos encontrar a los clientes con los valores mas bajos en la columna de "Recensy", es decir, se ubica
# a los clientes con la menor cantidad de dias desde la ultima compra

df_marketing.nsmallest(5, "Recency")

ID  Year_Birth   Education Marital_Status      Income   Kidhome  \
0   1826        1970  Graduation       Divorced  $84,835.00         0   
1      1        1961  Graduation         Single  $57,091.00         0   
2  10476        1958  Graduation        Married  $67,267.00         0   
3   1386        1967  Graduation       Together  $32,474.00         1   
4   5371        1989  Graduation         Single  $21,474.00         1   

   Teenhome Dt_Customer  Recency  MntWines  ...  NumStorePurchases  \
0         0     6/16/14        0       189  ...                  6   
1         0     6/15/14        0       464  ...                  7   
2         1     5/13/14        0       134  ...                  5   
3         1     5/11/14        0        10  ...                  2   
4         0      4/8/14        0         6  ...                  2   

   NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  \
0                  1             0             0             0             0   
1                  5             0             0             0             0   
2                  2             0             0             0             0   
3                  7             0             0             0             0   
4                  7             1             0             0             0   

   AcceptedCmp2  Response  Complain  Country  
0             0         1         0       SP  
1             1         1         0       CA  
2             0         0         0       US  
3             0         0         0      AUS  
4             0         1         0       SP  

[5 rows x 28 columns]

### Seleccion de columnas

In [39]:
# Vamos a seleccionar varias columnas del dataset

df_marketing[["Education"," Income ","Recency","Response"]].head()

Education      Income   Recency  Response
0  Graduation  $84,835.00         0         1
1  Graduation  $57,091.00         0         1
2  Graduation  $67,267.00         0         0
3  Graduation  $32,474.00         0         0
4  Graduation  $21,474.00         0         1

In [41]:
# Vamos a seleccionar una sola columna

df_marketing["Response"].head()

0    1
1    1
2    0
3    0
4    1
Name: Response, dtype: int64

In [42]:
# Otra forma de seleccionar una sola columna

df_marketing.Response.head()

0    1
1    1
2    0
3    0
4    1
Name: Response, dtype: int64

In [45]:
# Seleccionar columnas que inicien con una palabra, en este caso "Accepted"

df_marketing.filter(regex = "^Accepted").head()

AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2
0             0             0             0             0             0
1             0             0             0             0             1
2             0             0             0             0             0
3             0             0             0             0             0
4             1             0             0             0             0

In [46]:
# Seleccionar columnas que terminen con la palabra "home"

df_marketing.filter(regex = "home$").head()

Kidhome  Teenhome
0        0         0
1        0         0
2        0         1
3        1         1
4        1         0

In [51]:
# Seleccionar las columnas desde "Year_Birth" hasta "Kidhome" incluyendose
# Tambien se pueden seleccionar filas, en este caso de la 1 hasta la 6 sin incluir la 6

df_marketing.loc[1:6,"Year_Birth":"Kidhome"].head()

Year_Birth   Education Marital_Status      Income   Kidhome
1        1961  Graduation         Single  $57,091.00         0
2        1958  Graduation        Married  $67,267.00         0
3        1967  Graduation       Together  $32,474.00         1
4        1989  Graduation         Single  $21,474.00         1
5        1958         PhD         Single  $71,691.00         0

In [52]:
# Aqui seleccionamos las columnas 1, 2 y 5
# Seleccionamos todas las filas con :

df_marketing.iloc[:,[1,2,5]].head(6)

Year_Birth   Education  Kidhome
0        1970  Graduation        0
1        1961  Graduation        0
2        1958  Graduation        0
3        1967  Graduation        1
4        1989  Graduation        1
5        1958         PhD        0

### Summarize Data

In [4]:
# Veamos que nivel de escolaridad tienen los clientes registrados en la campaña de marketing

df_marketing["Education"].value_counts()

Graduation    1127
PhD            486
Master         370
2n Cycle       203
Basic           54
Name: Education, dtype: int64

In [5]:
# Con el siguiente codigo veremos el estado civil de los clientes

df_marketing["Marital_Status"].value_counts()

Married     864
Together    580
Single      480
Divorced    232
Widow        77
Alone         3
YOLO          2
Absurd        2
Name: Marital_Status, dtype: int64

In [13]:
# Aqui creamos 2 datasets. Uno con los clientes que aceptaron la oferta en la primera campaña y otra en la que no aceptaron

df_m1 = df_marketing[df_marketing.AcceptedCmp1 == 1]
df_m2 = df_marketing[df_marketing.AcceptedCmp1 == 0]

In [15]:
# En el siguiente codigo veremos la cantidad de quejas y el numero de niños en su hogar de los clientes 
# que si respondieron a la campaña de marqueting

print(df_m1.Complain.value_counts())
print(df_m1["Kidhome"].value_counts())

0    144
Name: Complain, dtype: int64
0    133
1      9
2      2
Name: Kidhome, dtype: int64


In [16]:
# Ahora veremos a los que no aceptaron la primera campaña

print(df_m2["Complain"].value_counts())
print(df_m2["Kidhome"].value_counts())

0    2075
1      21
Name: Complain, dtype: int64
0    1160
1     890
2      46
Name: Kidhome, dtype: int64


Podemos ver que las personas que no tienen niños en sus hogares y no han tenido quejas, son mas propensos a aceptar la oferta en la primera campaña de marketing

In [17]:
# Con la funcion .describe.() podemos ver las medidas de tendencia central y las medidas de dispersion de los datasets

df_m1.describe()

ID   Year_Birth     Kidhome    Teenhome     Recency  \
count    144.000000   144.000000  144.000000  144.000000  144.000000   
mean    5324.493056  1968.534722    0.090278    0.215278   46.979167   
std     3078.326355    13.530547    0.332677    0.445069   28.380647   
min      146.000000  1944.000000    0.000000    0.000000    1.000000   
25%     2793.750000  1957.000000    0.000000    0.000000   25.000000   
50%     4901.500000  1969.500000    0.000000    0.000000   45.500000   
75%     7896.750000  1979.000000    0.000000    0.000000   71.250000   
max    10965.000000  1994.000000    2.000000    2.000000   99.000000   

          MntWines   MntFruits  MntMeatProducts  MntFishProducts  \
count   144.000000  144.000000       144.000000       144.000000   
mean    758.604167   55.847222       433.638889        91.861111   
std     333.571922   55.342079       263.292935        73.293892   
min     152.000000    0.000000        21.000000         0.000000   
25%     502.000000   19.750000       196.000000        32.000000   
50%     761.500000   29.500000       437.500000        76.000000   
75%    1006.750000   87.750000       614.750000       146.250000   
max    1493.000000  190.000000       974.000000       253.000000   

       MntSweetProducts  ...  NumCatalogPurchases  NumStorePurchases  \
count        144.000000  ...           144.000000         144.000000   
mean          65.138889  ...             6.097222           8.062500   
std           54.960203  ...             2.517962           2.699958   
min            0.000000  ...             1.000000           4.000000   
25%           22.750000  ...             4.000000           6.000000   
50%           46.500000  ...             6.000000           8.000000   
75%           98.000000  ...             8.000000          10.000000   
max          194.000000  ...            11.000000          13.000000   

       NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  \
count         144.000000    144.000000    144.000000    144.000000   
mean            3.534722      0.166667      0.326389      0.472222   
std             2.044653      0.373979      0.470528      0.500970   
min             1.000000      0.000000      0.000000      0.000000   
25%             2.000000      0.000000      0.000000      0.000000   
50%             3.000000      0.000000      0.000000      0.000000   
75%             5.000000      0.000000      1.000000      1.000000   
max             9.000000      1.000000      1.000000      1.000000   

       AcceptedCmp1  AcceptedCmp2    Response  Complain  
count         144.0    144.000000  144.000000     144.0  
mean            1.0      0.090278    0.548611       0.0  
std             0.0      0.287580    0.499368       0.0  
min             1.0      0.000000    0.000000       0.0  
25%             1.0      0.000000    0.000000       0.0  
50%             1.0      0.000000    1.000000       0.0  
75%             1.0      0.000000    1.000000       0.0  
max             1.0      1.000000    1.000000       0.0  

[8 rows x 23 columns]

In [18]:
df_m2.describe()

ID   Year_Birth      Kidhome     Teenhome      Recency  \
count   2096.000000  2096.000000  2096.000000  2096.000000  2096.000000   
mean    5610.549141  1968.824427     0.468511     0.526240    49.255725   
std     3257.799795    11.873827     0.541333     0.545127    29.002896   
min        0.000000  1893.000000     0.000000     0.000000     0.000000   
25%     2834.750000  1959.750000     0.000000     0.000000    24.000000   
50%     5515.000000  1970.000000     0.000000     1.000000    50.000000   
75%     8447.500000  1977.000000     1.000000     1.000000    74.000000   
max    11191.000000  1996.000000     2.000000     2.000000    99.000000   

          MntWines    MntFruits  MntMeatProducts  MntFishProducts  \
count  2096.000000  2096.000000      2096.000000      2096.000000   
mean    272.698950    24.272424       148.627863        33.792462   
std     313.536172    37.649466       210.933538        51.048018   
min       0.000000     0.000000         0.000000         0.000000   
25%      21.000000     1.000000        14.000000         2.000000   
50%     152.000000     7.000000        57.000000        11.000000   
75%     448.000000    29.250000       197.500000        41.250000   
max    1486.000000   199.000000      1725.000000       259.000000   

       MntSweetProducts  ...  NumCatalogPurchases  NumStorePurchases  \
count       2096.000000  ...          2096.000000        2096.000000   
mean          24.447042  ...             2.426050           5.634065   
std           38.839822  ...             2.798614           3.227738   
min            0.000000  ...             0.000000           0.000000   
25%            1.000000  ...             0.000000           3.000000   
50%            7.000000  ...             1.000000           5.000000   
75%           29.000000  ...             4.000000           8.000000   
max          263.000000  ...            28.000000          13.000000   

       NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  \
count        2096.000000   2096.000000   2096.000000   2096.000000   
mean            5.438931      0.066317      0.057252      0.045324   
std             2.403079      0.248894      0.232379      0.208064   
min             0.000000      0.000000      0.000000      0.000000   
25%             4.000000      0.000000      0.000000      0.000000   
50%             6.000000      0.000000      0.000000      0.000000   
75%             7.000000      0.000000      0.000000      0.000000   
max            20.000000      1.000000      1.000000      1.000000   

       AcceptedCmp1  AcceptedCmp2     Response     Complain  
count        2096.0   2096.000000  2096.000000  2096.000000  
mean            0.0      0.008111     0.121660     0.010019  
std             0.0      0.089715     0.326971     0.099616  
min             0.0      0.000000     0.000000     0.000000  
25%             0.0      0.000000     0.000000     0.000000  
50%             0.0      0.000000     0.000000     0.000000  
75%             0.0      0.000000     0.000000     0.000000  
max             0.0      1.000000     1.000000     1.000000  

[8 rows x 23 columns]

Nuevamente pordemos concluir que las personas que no tienen reclamos y que no tienen hijos en sus hogares son mas propensos a aceptar la oferta en la primera campaña de marketing. Las personas que aceptan la oferta en la primera campaña suelen consumir mas vino, carne, frutas y dulces.

In [19]:
df_m1.shape , df_m2.shape

((144, 28), (2096, 28))

In [24]:
"La taza de conversion en la primera campaña es de: " + str((df_m1.shape[0] / (df_m1.shape[0] + df_m2.shape[0]))*100) + " %"

'La taza de conversion en la primera campaña es de: 6.428571428571428 %'

Quiere decir que del 100% de todos los clientes a los que se le ofrecio la primera campaña de marketing, solo tuvieron una respuesta positiva el 6.43%

### Summarize Data parte 2
Compararemos clientes que aceptaron la oferta en la primera campaña con los que la aceptaron en la sexta y ultima campaña

In [3]:
firstc = df_marketing[df_marketing["AcceptedCmp1"] == 1]
lastc = df_marketing[df_marketing["Response"] == 1]
print(firstc.shape, lastc.shape)

(144, 28) (334, 28)


In [9]:
# Creamos otro objeto vara ver la taza de convercion

tc1 = (firstc.shape[0] / df_marketing.shape[0])*100

In [12]:
print("La taza de convercion en la primera campaña fue de " + str(tc1) + "%")

La taza de convercion en la primera campaña fue de 6.428571428571428%


In [13]:
# Calculamos el promedio del primer dataset

firstc.mean()

C:\Users\MOISES~1\AppData\Local\Temp/ipykernel_14236/4240607091.py:3: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  firstc.mean()


ID                     5324.493056
Year_Birth             1968.534722
Kidhome                   0.090278
Teenhome                  0.215278
Recency                  46.979167
MntWines                758.604167
MntFruits                55.847222
MntMeatProducts         433.638889
MntFishProducts          91.861111
MntSweetProducts         65.138889
MntGoldProds             77.131944
NumDealsPurchases         1.416667
NumWebPurchases           5.729167
NumCatalogPurchases       6.097222
NumStorePurchases         8.062500
NumWebVisitsMonth         3.534722
AcceptedCmp3              0.166667
AcceptedCmp4              0.326389
AcceptedCmp5              0.472222
AcceptedCmp1              1.000000
AcceptedCmp2              0.090278
Response                  0.548611
Complain                  0.000000
dtype: float64

In [14]:
lastc.mean()

C:\Users\MOISES~1\AppData\Local\Temp/ipykernel_14236/1022178384.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  lastc.mean()


ID                     5421.820359
Year_Birth             1969.416168
Kidhome                   0.341317
Teenhome                  0.305389
Recency                  35.383234
MntWines                502.703593
MntFruits                38.203593
MntMeatProducts         294.353293
MntFishProducts          52.050898
MntSweetProducts         38.634731
MntGoldProds             61.446108
NumDealsPurchases         2.335329
NumWebPurchases           5.071856
NumCatalogPurchases       4.203593
NumStorePurchases         6.095808
NumWebVisitsMonth         5.293413
AcceptedCmp3              0.230539
AcceptedCmp4              0.185629
AcceptedCmp5              0.275449
AcceptedCmp1              0.236527
AcceptedCmp2              0.059880
Response                  1.000000
Complain                  0.008982
dtype: float64

¿Quienes son mas jovenes?

In [16]:
print(firstc.mean()[1], lastc.mean()[1])

1968.5347222222222 1969.4161676646706


C:\Users\MOISES~1\AppData\Local\Temp/ipykernel_14236/558685792.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  print(firstc.mean()[1], lastc.mean()[1])


No existe una diferencia significativa, pero los que compraron en la ultima campaña eran un poco mas jovenes

¿Que cantidad de niños tienen en sus hogares los 2 grupos de clientes?

In [17]:
print(firstc.mean()[2], lastc.mean()[2])

0.09027777777777778 0.3413173652694611


C:\Users\MOISES~1\AppData\Local\Temp/ipykernel_14236/1436301874.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  print(firstc.mean()[2], lastc.mean()[2])


Los clientes que respondieron en la primera campaña tienen mas niños que los de la segunda campaña

Numero de dias desde la ultima compra

In [18]:
print(firstc.mean()[4], lastc.mean()[4])

46.979166666666664 35.383233532934135


C:\Users\MOISES~1\AppData\Local\Temp/ipykernel_14236/2541765040.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  print(firstc.mean()[4], lastc.mean()[4])


Los clientes que compraron en la ultima campaña tienen menos dias de haber realizado la ultima compra

¿Quienes consumen mas vino?

In [19]:
print(firstc.mean()[5], lastc.mean()[5])

758.6041666666666 502.70359281437123


C:\Users\MOISES~1\AppData\Local\Temp/ipykernel_14236/697525793.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  print(firstc.mean()[5], lastc.mean()[5])


Los clientes que respondieron a la primera campaña consumen mas vino

¿Que grupo de clientes ha tenido mas reclamos?

In [5]:
print(firstc.mean()[-1], lastc.mean()[-1])

0.0 0.008982035928143712


C:\Users\MOISES~1\AppData\Local\Temp/ipykernel_7908/156532798.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  print(firstc.mean()[-1], lastc.mean()[-1])


Los clientes que respondieron a la primera campaña no tienen reclamos

### Funcion Merge (unir)

In [2]:
# Vamos a crear un dataframe (Primera forma)

staff_df = pd.DataFrame({'Name':['Kelly','Sally','James'],
                         'Role':['Director of RH','Course Liasion','Grader']})
staff_df

Name            Role
0  Kelly  Director of RH
1  Sally  Course Liasion
2  James          Grader

In [3]:
# La funcion .set_index() se utiliza para establecer columnas como indice del dataframe

staff_df = staff_df.set_index('Name')
staff_df

Role
Name                 
Kelly  Director of RH
Sally  Course Liasion
James          Grader

In [14]:
# Creamos el segundo dataframe (Segunda forma)

student_df = pd.DataFrame([{'Name':'James','School':'Business'},
                          {'Name':'Myke','School':'Law'},
                          {'Name':'Sally','School':'Engineering'}])
student_df

Name       School
0  James     Business
1   Myke          Law
2  Sally  Engineering

In [15]:
student_df = student_df.set_index('Name')
student_df

School
Name              
James     Business
Myke           Law
Sally  Engineering

In [17]:
print(staff_df)
print(student_df)

                 Role
Name                 
Kelly  Director of RH
Sally  Course Liasion
James          Grader
            School
Name              
James     Business
Myke           Law
Sally  Engineering


### Outer
Aqui obtenemos un dataset con todos los registros de los dos datasets

In [20]:
pd.merge(staff_df, student_df, how = 'outer', left_index = True, right_index = True)

Role       School
Name                              
James          Grader     Business
Kelly  Director of RH          NaN
Myke              NaN          Law
Sally  Course Liasion  Engineering

### Inner
Aqui obtenemos un dataset con los registros en comun de los dos datasets

In [28]:
pd.merge(staff_df, student_df, how = 'inner', left_index = True, right_index = True)

Role       School
Name                              
Sally  Course Liasion  Engineering
James          Grader     Business

### Left
Aqui obtenemos un dataset con los registros del dataset izquierdo y con la informacion de las columnas correspondientes al dataset derecho

In [29]:
pd.merge(staff_df, student_df, how = 'left', left_index = True, right_index = True)

Role       School
Name                              
Kelly  Director of RH          NaN
Sally  Course Liasion  Engineering
James          Grader     Business

### Right
Aqui obtenemos un dataset con los registros del dataset derecho y con la informacion de las columnas correspondientes al dataset izquierdo

In [30]:
pd.merge(staff_df, student_df, how = 'right', left_index = True, right_index = True)

Role       School
Name                              
James          Grader     Business
Myke              NaN          Law
Sally  Course Liasion  Engineering

### Right sin index

In [31]:
# Aqui vamos a eliminar los indices asignados previamente en los dos datasets

staff_df = staff_df.reset_index()
student_df = student_df.reset_index()

In [34]:
# Aqui obtenemos un dataset con los registros del dataset derecho y con la informacion de las columnas correspondientes
# al dataser izquierdo

pd.merge(staff_df, student_df, how = 'right', on = 'Name')

Name            Role       School
0  James          Grader     Business
1   Myke             NaN          Law
2  Sally  Course Liasion  Engineering

### Funcion Merge segunda parte

In [35]:
staff_df_2 = pd.DataFrame({'Name':['Kelly','Sally','James'],
                          'Role':['Director o HR','Course Liation','Grader'],
                          'Location':['State Street','Washingtong Avenue','Washington Avenue']})

student_df_2 = pd.DataFrame({'Name':['James','Mike','Sally'],
                            'School':['Business','Law','Engineering'],
                            'Location':['1024 Billiard Avenue','Fraternity House #22','512 Wilson Crescent']})

In [36]:
staff_df_2

Name            Role            Location
0  Kelly   Director o HR        State Street
1  Sally  Course Liation  Washingtong Avenue
2  James          Grader   Washington Avenue

### Left sin index

In [37]:
# Aqui obtenemos un dataset con los registros del dataset izquierdo y con la informacion de las columnas correspondientes
# al dataser derecho

pd.merge(staff_df_2, student_df_2, how = 'left', on = 'Name')

Name            Role          Location_x       School  \
0  Kelly   Director o HR        State Street          NaN   
1  Sally  Course Liation  Washingtong Avenue  Engineering   
2  James          Grader   Washington Avenue     Business   

             Location_y  
0                   NaN  
1   512 Wilson Crescent  
2  1024 Billiard Avenue

### Creacion de dataset 3

In [38]:
staff_df_3 = pd.DataFrame({'First Name':['Kelly','Sally','James'],
                           'Last Name':['Desjardins','Brooks','Wilde'],
                           'Role':['Director o HR','Course Liation','Grader']})

student_df_3 = pd.DataFrame({'First Name':['James','Mike','Sally'],
                             'Last Name':['Hammond','Smith','Brooks'],
                             'School':['Business','Law','Engineering']})

In [39]:
student_df_3

First Name Last Name       School
0      James   Hammond     Business
1       Mike     Smith          Law
2      Sally    Brooks  Engineering

### Inner con lista

In [41]:
# Aqui creamos un dataframe donde se ubican los registros exactos segun la lista.

pd.merge(staff_df_3, student_df_3, how = 'inner', on = ['First Name', 'Last Name'])

First Name Last Name            Role       School
0      Sally    Brooks  Course Liation  Engineering

Podemos observar que el unico registro es Sally Brooks porque es el unico que coniside con  'First Name' y 'Last Name'